generator/_render.py

```diff
@@ -0,0 +1,288 @@
+    ordered dict mapping method names to method named tuples, like:
+        {
+            'client': {
+                'watch': (
+                    name='watch',
+                    link='watchcallback',
+                    doc='## watch(callback) ...',
+                ),
+                'foo': (
+                    name='foo',
+                    link='fooargs-callback',
+                    doc='## foo(args, callback) ...',
+                ),
+            },
+        }
+    """
+    template = get_template('wrappers.md')
+    outfile = os.path.join(out, 'wrappers.md')
+    template.stream().dump(outfile)
+    with open(outfile) as stream:
+        content = stream.read()
+    wrappers_info = {}
+    method = collections.namedtuple('Method', 'name link doc')
+    for part in content.split('\n## '):
+        if part.startswith('#'):
+            continue
+        facade_name, method_doc = part.split('.', 1)
+        method_name = method_doc.split('(', 1)[0]
+        method_link = method_doc.split('\n', 1)[0].replace(' ', '-')
+        for char in '(),':
+            method_link = method_link.replace(char, '')
+        methods = wrappers_info.setdefault(
+            facade_name, collections.OrderedDict())
+        methods[method_name] = method(
+            name=method_name, link=method_link, doc='## ' + method_doc)
+    return wrappers_info
+
+
+def _handle_methods(name, methods, types):
+    if not isinstance(methods, list):
+        raise TypeError('methods for facade {} is not a list'.format(name))
+    resulting_methods = []
+    for method in methods:
+        if not isinstance(method, collections.Mapping):
+            raise TypeError('method for facade {} is not a map'.format(name))
+        try:
+            method_name = method['Name']
+        except KeyError:
+            raise ValueError('method for facade {} has no name'.format(name))
+        try:
+            params = from_gotype(types, method.get('Param'))
+            result = from_gotype(types, method.get('Result'))
+        except (KeyError, ValueError) as err:
+            raise ValueError(
+                'method {}.{} is not valid: {}'.format(name, method_name, err))
+        doc = method.get('Doc', 'There is no documentation for this method.')
+        resulting_methods.append(Method(
+            request=method_name,
+            params=params,
+            result=result,
+            doc=wrap_text(doc, indent='    ', trailing_newline=True)))
+    # Resulting methods are sorted mostly for making tests deterministic.
+    return sorted(resulting_methods)
+
+
+def _hyphenize(string):
+    """Convert thisString into this-string."""
+    parts = []
+    for char in string:
+        if char.isupper():
+            char = '-' + char.lower()
+        parts.append(char)
+    return ''.join(parts)
+
+
+def _nowstr():
+    """Return the current time in UTC as a string."""
+    return datetime.utcnow().strftime('%a %Y/%m/%d %H:%M:%S UTC')
```

Comment:
Maybe this time should be cached so it's a build time instead of each time it's called returning a different time. Even if that is just seconds.... 🤷‍♂️

Reply:
I think actually since it's to show when the file is generated exact timestamp is preferred.

Reply:
Yes, it's meant to be called every time, and arguably it's very cheap.